File: javascript/BinaryHeap.js
class MaxBinaryHeap{
    constructor(){
        this.values = []
    }
    insert(value){
        this.values.push(value)
        this.puppleUp()
        return this.values
    }
    puppleUp(){
        let idx = this.values.length - 1
        let element = this.values[idx]
        let parentIdx, parent;
        while(idx > 0){
            parentIdx = Math.floor((idx - 1)/2)
            parent = this.values[parentIdx]
            if (element <= parent) break
            this.values[parentIdx] = element
            this.values[idx] = parent
            idx = parentIdx   
        }
    }
    extractMax(){
        let max = this.values[0]
        let end = this.values.pop()
        //  edge case if length > 0 [puppleDown]
        if (this.values.length > 0){
            this.values[0] = end
            this.puppleDown()   
        }
        return max
    }
    puppleDown(){
        let Idx = 0;
        let length = this.values.length
        let element = this.values[Idx]
        let leftChildIdx, rightChildIdx, leftChild, rightChild, swapIdx;
        while(true){
            leftChildIdx = 2 * Idx + 1
            rightChildIdx = 2 * Idx + 2
            swapIdx = null
            if (leftChildIdx < length){
                leftChild = this.values[leftChildIdx]
                if (leftChild > element){
                    swapIdx = leftChildIdx
                }
            }

            if (rightChildIdx < length){
                rightChild = this.values[rightChildIdx]
                if (
                    (swapIdx === null && rightChild > element) || 
                    (swapIdx !== null && rightChild > leftChild)
                    ){
                    swapIdx = rightChildIdx
                }
            }
            if(swapIdx === null)break;
            this.values[Idx] = this.values[swapIdx];
            this.values[swapIdx] = element;
            Idx = swapIdx;
        }
    }
}

// let maxBinaryHeap = new MaxBinaryHeap()
// maxBinaryHeap.insert(50)
// maxBinaryHeap.insert(40)
// maxBinaryHeap.insert(60)
// maxBinaryHeap.insert(30)
// maxBinaryHeap.insert(70)
// maxBinaryHeap.insert(20)
// maxBinaryHeap.insert(80)
// maxBinaryHeap.insert(10)


class MinBinaryHeap{
    constructor(){
        this.values = []
    }

    insert(value){
        this.values.push(value)
        this.puppleUp()   
        return this.values
    }

    puppleUp(){
        let idx = this.values.length - 1 
        let element = this.values[idx]
        let parentIdx;
        let parent;
        while(idx > 0){
            parentIdx = Math.floor((idx -1)/2)
            parent = this.values[parentIdx]
            if (element >= parent) break
            this.values[parentIdx] = element
            this.values[idx] = parent
            idx = parentIdx
        }
    }

    extractMin(){
        let min = this.values[0]
        let end = this.values.pop()
        //  edge case if length > 0 [puppleDown]
        if (this.values.length > 0){
            this.values[0] = end
            this.puppleDown()
        }
        return min
    }
    puppleDown(){
        let idx = 0;
        let element = this.values[0]
        let length = this.values.length
        let leftChildIdx, rightChildIdx, leftChild, rightChild, swapIdx;
        while(true){
            leftChildIdx = 2 * idx + 1
            rightChildIdx = 2 * idx + 2
            swapIdx = null

            if (leftChildIdx < length){
                leftChild = this.values[leftChildIdx]
                if (leftChild < element){
                    swapIdx = leftChildIdx
                } 
            }
            if (rightChildIdx < length){
                rightChild = this.values[rightChildIdx]
                if (
                    (swapIdx === null && rightChild < element ) || 
                    (swapIdx !== null && rightChild < leftChild)
                    )
                    swapIdx = rightChildIdx
            }

            if (swapIdx === null) break
            this.values[idx] = this.values[swapIdx]
            this.values[swapIdx] = element
            idx = swapIdx
        }
    }
}

let minBinaryHeap = new MinBinaryHeap()
minBinaryHeap.insert(50)
minBinaryHeap.insert(40)
minBinaryHeap.insert(60)
minBinaryHeap.insert(30)
minBinaryHeap.insert(70)
minBinaryHeap.insert(20)
minBinaryHeap.insert(80)
minBinaryHeap.insert(10)

File: python/BinaryHeap.py
import math
class MaxBinaryHeap:
    def __init__(self):
        self.values = []

    def insert(self, value):
        self.values.append(value)
        self.bubble_up()

    def get_values(self):
        return self.values

    def bubble_up(self):
        idx = len(self.values) - 1
        element = self.values[idx]
        parentIdx = None
        parent = None

        while idx > 0 : # edge case for idx > 0 to eliminate undefined (comparing with negative index)
            parentIdx = math.floor((idx - 1)/2)
            parent = self.values[parentIdx]

            if element <= parent: break
            self.values[idx] = parent
            self.values[parentIdx] = element
            idx = parentIdx

    def extract_max(self):
        idx = 0
        max = self.values[0]
        end = self.values.pop()
        #  edge case if len > 1
        if len(self.values) > 1:
            self.values[0] = end 
            self.pupple_down()
        return max
        
    def pupple_down(self):
        idx = 0
        element = self.values[0]
        length = len(self.values)
        left_child_idx = None
        right_child_idx = None
        left_child = None
        right_child = None
        swap_idx = None

        while True:
            left_child_idx = 2 * idx  + 1
            right_child_idx = 2 * idx  + 2
            swap_idx = None

            if left_child_idx < length:
                left_child = self.values[left_child_idx]
                if left_child > element:
                    swap_idx = left_child_idx
            
            if right_child_idx < length:
                right_child = self.values[right_child_idx]
                if (swap_idx == None and right_child > element) or (swap_idx is not None and right_child > left_child):
                    swap_idx = right_child_idx

            if swap_idx is None: break
            self.values[idx] = self.values[swap_idx]
            self.values[swap_idx] = element
            idx = swap_idx



# max_binary_heap = MaxBinaryHeap()
# max_binary_heap.insert(50)
# max_binary_heap.insert(40)
# max_binary_heap.insert(60)
# max_binary_heap.insert(30)
# max_binary_heap.insert(70)
# max_binary_heap.insert(20)
# max_binary_heap.insert(80)
# max_binary_heap.insert(10)
# max_binary_heap.insert(90)
# max_binary_heap.insert(110)
# max_binary_heap.extract_max()

# values = max_binary_heap.get_values()
# print(values)
# max = max_binary_heap.extract_max()
# print(max)
# max = max_binary_heap.extract_max()
# print(max)
# print(max)
# max = max_binary_heap.extract_max()
# print(max)
# values = max_binary_heap.get_values()
# print(values)
